packages/headless/src/lib/headless.service.ts

```diff
@@ -0,0 +1,531 @@
+  }
+
+  public fetchUserPreferences({
+    listener,
+    onSuccess,
+    onError,
+  }: {
+    listener: (result: FetchResult<IUserPreferenceSettings[]>) => void;
+    onSuccess?: (messages: IUserPreferenceSettings[]) => void;
+    onError?: (error: unknown) => void;
+  }) {
+    this.assertSessionInitialized();
+
+    const { unsubscribe } = this.queryService.subscribeQuery({
+      options: {
+        ...this.userPreferencesQueryOptions,
+        onSuccess,
+        onError,
+      },
+      listener: (result) => this.callFetchListener(result, listener),
+    });
+
+    return unsubscribe;
+  }
+
+  public async updateUserPreferences({
+    templateId,
+    channelType,
+    checked,
+    listener,
+    onSuccess,
+    onError,
+  }: {
+    templateId: IUpdateUserPreferencesVariables['templateId'];
+    channelType: IUpdateUserPreferencesVariables['channelType'];
+    checked: IUpdateUserPreferencesVariables['checked'];
+    listener: (
+      result: UpdateResult<
+        IUserPreferenceSettings,
+        unknown,
+        IUpdateUserPreferencesVariables
+      >
+    ) => void;
+    onSuccess?: (settings: IUserPreferenceSettings) => void;
+    onError?: (error: unknown) => void;
+  }) {
+    this.assertSessionInitialized();
+
+    const { result, unsubscribe } = this.queryService.subscribeMutation<
+      IUserPreferenceSettings,
+      unknown,
+      IUpdateUserPreferencesVariables
+    >({
+      options: {
+        mutationFn: (variables) =>
+          this.api.updateSubscriberPreference(
+            variables.templateId,
+            variables.channelType,
+            variables.checked
+          ),
+        onSuccess: (data) => {
+          this.queryClient.setQueryData<IUserPreferenceSettings[]>(
+            USER_PREFERENCES_QUERY_KEY,
+            (oldUserPreferences) =>
+              oldUserPreferences.map((setting) => {
+                if (setting.template._id === data.template._id) {
+                  return data;
+                }
+
+                return setting;
+              })
+          );
+        },
+      },
+      listener: (res) => this.callUpdateListener(res, listener),
+    });
+
+    result
+      .mutate({ templateId, channelType, checked })
+      .then((data) => {
+        onSuccess?.(data);
+
+        return data;
+      })
+      .catch((error) => {
+        onError?.(error);
+      })
+      .finally(() => {
+        unsubscribe();
+      });
+  }
+
+  public async markNotificationsAsRead({
+    messageId,
+    listener,
+    onSuccess,
+    onError,
+  }: {
+    messageId: string | string[];
+    listener: (
+      result: UpdateResult<IMessage, unknown, { messageId: string | string[] }>
+    ) => void;
+    onSuccess?: (message: IMessage) => void;
+    onError?: (error: unknown) => void;
+  }) {
+    this.assertSessionInitialized();
+
+    const { result, unsubscribe } = this.queryService.subscribeMutation<
+      IMessage,
+      unknown,
+      { messageId: string | string[] }
```

Comment:
It is unfortunate that due to our API implementation this typing has to creep into your new package. But I'd suggest here to just implement `string[]` as `messageId` type because that can hold `[messageId1]` as equivalent of `string` type. And this will allow in the future to simplify the signature of the API endpoint.

Reply:
I'm sorry but I'm not getting what you mean. Can you show by example?